[clang][SPIR-V] Always add convergence intrinsics

Keenuts opened the pull request:
PR #80680 added bits in the codegen to lazily add convergence intrinsics when required. This logic relied on the LoopStack. The issue is when parsing the condition, the loopstack doesn't yet reflect the correct values, as expected since we are not yet in the loop.

However, convergence tokens should sometimes already be available. The solution which seemed the simplest is to greedily generate the tokens when we generate SPIR-V.

Fixes #88144

Keenuts committed:
[clang][SPIR-V] Always add convervence intrinsics

PR #80680 added bits in the codegen to lazily add convergence intrinsics
when required. This logic relied on the LoopStack. The issue is
when parsing the condition, the loopstack doesn't yet reflect the
correct values, as expected since we are not yet in the loop.

However, convergence tokens should sometimes already be available.
The solution which seemed the simplest is to greedily generate the
tokens when we generate SPIR-V.

Fixes #88144

Signed-off-by: Nathan Gauër <brioche@google.com>

arsenm commented on clang/lib/CodeGen/CGCall.cpp:
```diff
@@ -4824,6 +4824,9 @@ llvm::CallInst *CodeGenFunction::EmitRuntimeCall(llvm::FunctionCallee callee,
   call->setCallingConv(getRuntimeCC());
+
+  if (getTarget().getTriple().isSPIRVLogical() && call->isConvergent())
+    return dyn_cast<llvm::CallInst>(addControlledConvergenceToken(call));
```
can this just be cast<>?

arsenm commented on clang/lib/CodeGen/CodeGenFunction.h:
```diff
@@ -4987,7 +4990,11 @@ class CodeGenFunction : public CodeGenTypeCache {
-  llvm::CallBase *addControlledConvergenceToken(llvm::CallBase *Input);
+  template <typename CallType>
+  CallType *addControlledConvergenceToken(CallType *Input) {
+    return dyn_cast<CallType>(
```
should just be cast? You also have yet another dyn_cast on the use?

Keenuts replied:
Forgot to fix the use, and yes, should be a cast. Bad habits...

Keenuts committed:
feedback review

arsenm commented on clang/lib/CodeGen/CGCall.cpp:
```diff
@@ -4824,6 +4824,9 @@ llvm::CallInst *CodeGenFunction::EmitRuntimeCall(llvm::FunctionCallee callee,
       callee, args, getBundlesForFunclet(callee.getCallee()), name);
   call->setCallingConv(getRuntimeCC());
+
+  if (getTarget().getTriple().isSPIRVLogical() && call->isConvergent())
```
Shouldn't have this target check

arsenm commented on clang/lib/CodeGen/CGStmt.cpp:
```diff
@@ -1109,6 +1124,10 @@ void CodeGenFunction::EmitForStmt(const ForStmt &S,
   llvm::BasicBlock *CondBlock = CondDest.getBlock();
   EmitBlock(CondBlock);
 
+  if (getTarget().getTriple().isSPIRVLogical())
```
None of these target checks have anything to do with the target. If you want to conditionally enable these for bringup, these should be hidden in an enableConvergenceTokenEmission or something predicate

Keenuts replied:
Right, something at the module level, which hides this into a more genering function, like the `shouldEmitRTTI`. Will fix.

Keenuts replied:
Changed those for a module-level function which hides the target specific bit behind a more generic check.
Let me know if that's not what you had in mind!

Keenuts committed:
review feedback target check

Signed-off-by: Nathan Gauër <brioche@google.com>

bogner commented on clang/lib/CodeGen/CGStmt.cpp:
```diff
@@ -3101,3 +3130,68 @@ CodeGenFunction::GenerateCapturedStmtFunction(const CapturedStmt &S) {
+  if (BB->empty())
+    Builder.SetInsertPoint(BB);
+  else
+    Builder.SetInsertPoint(&BB->front());
```
Should this be something like `BB->getFirstInsertionPt()`? That's more concise and better behaved around PHI nodes and the like

bogner commented on clang/lib/CodeGen/CGStmt.cpp:
```diff
@@ -3101,3 +3130,68 @@ CodeGenFunction::GenerateCapturedStmtFunction(const CapturedStmt &S) {
+llvm::IntrinsicInst *
+CodeGenFunction::getOrEmitConvergenceEntryToken(llvm::Function *F) {
+  auto *BB = &F->getEntryBlock();
+  auto *token = getConvergenceToken(BB);
```
LLVM naming prefers `Token`

llvm-beanz commented on clang/lib/CodeGen/CGStmt.cpp:
```diff
@@ -3101,3 +3130,68 @@ CodeGenFunction::GenerateCapturedStmtFunction(const CapturedStmt &S) {
+      llvm::Intrinsic::experimental_convergence_loop, {}, {});
+  Builder.restoreIP(IP);
+
+  auto I = addConvergenceControlToken(CB, ParentToken);
```
I'm pretty sure that some of this is just code that you moved from CGBuiltin.cpp, but we probably should change the gratuitous use of `auto`. LLVM's conventions only use `auto` when it improves readability (see: https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable).

In quite a few of these cases the explicit type would be much more readable.

Keenuts replied:
Right, replaced the auto usage

Keenuts committed:
review feedback

Keenuts committed:
fix changing register name in tests

Signed-off-by: Nathan Gauër <brioche@google.com>

Keenuts commented:
Hi all, rebased on main, and addressed the comments.
 This commits changes the register order on SPIR-V vs DXIL, which required me to fix the mad+lerp intrinsic tests. Should be NFC, just storing the register name in a CHECK variable.

arsenm commented on clang/lib/CodeGen/CodeGenModule.h:
```diff
@@ -1586,6 +1586,12 @@ class CodeGenModule : public CodeGenTypeCache {
+  // Return whether structured convergence intrinsics should be generated for
+  // this target.
+  bool shouldEmitConvergenceTokens() const {
+    return getTriple().isSPIRVLogical();
```
Should add a TODO this should just be unconditional in the future

Keenuts replied:
Not necessarily. The SPIR-V compute (as OpenCL not Vulkan compute) doesn't require a structured CFG, so this could remain forever if they don't see any benefit in generating a SCFG.

arsenm replied:
This doesn't have anything to do with what the target wants to do with with the CFG lowering. Structurally the IR should not allow uncontrolled convergence, and that it exists is a wart that needs to exist until everywhere handles convergence tokens

Keenuts replied:
That makes sense. Added a TODO line!

Keenuts commented:
Thanks for the reviews. Waiting for 1 approval from MS and I'll merge

Keenuts committed:
add TODO comment

Keenuts commented:
rebased on main, local tests are passing, waiting on CI to merge.
